src/example-types/dateRangeFacet.js

```diff
@@ -0,0 +1,85 @@
+let F = require('futil')
+let _ = require('lodash/fp')
+let { getDateIfValid, rollingRangeToDates } = require('../dateUtil')
+
+let getDateRange = (range, timezone) => {
+  let { from, to } = rollingRangeToDates(range, timezone)
+  return F.compactObject({
+    from: getDateIfValid(from),
+    to: getDateIfValid(to),
+  })
+}
+
+module.exports = {
+  hasValue: context => _.get('values.length', context),
+  /**
+   * VALID CONTEXT
+   * 1. Must have "field" (String) and "ranges" (Array) properties
+   * 2. Each range object must have a "key" property
+   * 3. Each range must have a "range" property with the range phrase
+   */
+  validContext: context =>
+    _.has('field', context) &&
+    !!_.get('ranges.length', context) &&
+    _.every(r => _.has('key', r) && _.has('range', r), context.ranges),
+  /**
+   * FILTER
+   * Based on the keys checked we get the actual values
+   * from the context.ranges and compose a bool/should query.
+   **/
+  filter(context) {
+    let { field, ranges, values, timezone = 'UTC' } = context
+    let should = _.flow(
+      _.filter(r => _.includes(r.key, values)),
+      _.map(({ range }) => ({
+        range: {
+          [field]: {
+            format: 'dateOptionalTime',
+            ...getDateRange(range, timezone),
+          },
+        },
+      }))
+    )(ranges)
+
+    return { bool: { should } }
+  },
+  /**
+   * RESULT
+   * Context should have `ranges` prop where each range has a "key"
+   * Ranges should have 'range' prop containing the range phrase (eg. 'allFutureDates')
+   * Example:
+      ranges: [
+        { range: 'allFutureDates', key: "open" },
+        { range: 'allPastDates', key: "expired" }
+      ]
+   */
+  result(context, search) {
+    let { field, format, timezone, ranges } = context
+    return search({
+      aggs: {
+        range: {
+          date_range: {
+            field,
+            format: format || "yyyy-MM-dd'T'HH:mm:ss.SSSZ",
+            ranges: _.map(
+              ({ range, key }) => ({
+                key,
+                ...getDateRange(range, timezone),
```

Comment:
The `to` value will be excluded, which in a number of the datemath expressions includes `-1ms`. You may want to remove the `-1ms` in order to be more precise. See: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-daterange-aggregation.html#search-aggregations-bucket-daterange-aggregation

Reply:
For the date range based on the range phrase in some cases `from` or `to` would be excluded (e.g. `allPastDates` and `allFutureDates`). And that is fine. For example, in the case of `allFutureDates` the `to` would be removed and only the `from` should be there. `to` in this case is simply an empty string and the `date_range/ranges` `agg` does not like empty strings due to the format specified. 

The values which are not empty pass through the `ES dateMath` lib and get converted to dates based on the dateMath specified in the `rangeToDatemath` (which already have the `-1ms`). Maybe I am missing something but I am not seeing the issue here ...

Reply:
It's likely not an issue since we are talking about such a small amount of time, but since ES is always doing a less than on the `to` for the date range the `-1ms` is unneeded.